addons/sale/models/sale_order.py

```diff
@@ -801,6 +803,65 @@ def _onchange_prepayment_percent(self):
         if not self.prepayment_percent:
             self.require_payment = False
 
+    @api.onchange('order_line')
+    def _onchange_order_line(self):
+        virtual_ids = self.order_line.mapped('virtual_id')
+        lines_to_delete = self.order_line.filtered(
+            lambda sol: bool(
+                # Lines which aren't saved in DB yet.
+                virtual_ids and sol.linked_virtual_id and sol.linked_virtual_id not in virtual_ids
+            ) or bool(
+                # Lines which are saved in DB.
+                sol.linked_line_id and sol.linked_line_id not in self.order_line
+            )
+        )
+        if lines_to_delete:
+            self.order_line = [Command.unlink(line.id) for line in lines_to_delete]
```

Comment:
@Feyensv Why `unlink` and not `delete`?

Reply:
I think both work the same here, but yeah, delete is better indeed :ok_hand: